c2rust/src/main.rs

```diff
@@ -1,56 +1,154 @@
+                continue;
+            }
+            let file_name = entry.file_name();
+            let file_name: &Utf8Path = Path::new(&file_name).try_into()?;
+            let file_name = file_name.as_str();
+            let name = match file_name.strip_prefix(c2rust_name) {
+                Some(name) => name,
+                None => continue,
+            };
+            let name = match name.strip_prefix('-') {
+                Some(name) => name,
+                None => continue,
+            };
+            let path = entry.path();
+            if !path.is_executable() {
+                continue;
+            }
+            sub_commands.push(Self {
+                path: Some(path),
+                name: name.to_owned().into(),
+            });
         }
-    };
+        Ok(sub_commands)
+    }
+
+    /// Get all known [`SubCommand`]s.  These have no [`SubCommand::path`].
+    /// Even if the subcommand executables aren't there, we can still suggest them.
+    pub fn known() -> impl Iterator<Item = Self> {
+        ["transpile", "instrument", "pdg", "analyze"]
+            .into_iter()
+            .map(|name| Self {
+                path: None,
+                name: name.into(),
+            })
+    }
+
+    /// Get all known ([`Self::known`]) and actual, found ([`Self::find_all`]) subcommands,
+    /// putting the known ones first so that the found ones overwrite them and take precedence.
+    pub fn all() -> anyhow::Result<impl Iterator<Item = Self>> {
+        Ok(Self::known().chain(Self::find_all()?))
+    }
+
+    pub fn invoke<I, S>(&self, args: I) -> anyhow::Result<()>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let path = self.path.as_ref().ok_or_else(|| {
+            anyhow!(
+                "known subcommand not found (probably not built): {}",
+                self.name
+            )
+        })?;
+        let status = Command::new(&path).args(args).status()?;
+        process::exit(status.code().unwrap_or(1));
+    }
 }
 
-fn invoke_subcommand<I, S>(subcommand: &str, args: I)
-where
-    I: IntoIterator<Item = S>,
-    S: AsRef<OsStr>,
-{
-    // Assumes the subcommand executable is in the same directory as this driver
-    // program.
-    let cmd_path = std::env::current_exe().expect("Cannot get current executable path");
-    let mut cmd_path = cmd_path.as_path().canonicalize().unwrap();
-    cmd_path.pop(); // remove current executable
-    cmd_path.push(format!("c2rust-{}", subcommand));
-    assert!(cmd_path.exists(), "{:?} is missing", cmd_path);
-    exit(
-        Command::new(cmd_path.into_os_string())
-            .args(args)
-            .status()
-            .expect("SubCommand failed to start")
-            .code()
-            .unwrap_or(-1),
-    );
+fn main() -> anyhow::Result<()> {
+    let sub_commands = SubCommand::all()?.collect::<Vec<_>>();
+    let sub_commands = sub_commands
+        .iter()
```

Comment:
Why make a vec then immediately iterate? Can't you just do `all()?.map(...)`?

Reply:
No they have to be stored somewhere.  Afterwards they're references to the ones in the `Vec`.  I'm pretty sure I tried your suggestion already and it yelled at me.

Reply:
Hm, I figured the HashMap would do that, but I guess if you're storing the command name as a ref then that won't work. Maybe store the key as a String?

Reply:
I think it's just simpler this way.  Other ways I was running into self-referential data.

Reply:
Fair enough. It's a small thing anyway.